code/modules/projectiles/attachments/laser_sight.dm

```diff
@@ -0,0 +1,140 @@
+/obj/item/attachment/laser_sight
+	name = "laser sight"
+	desc = "A glorified laser pointer. Good for knowing what you're aiming at."
+	icon_state = "laser_sight"
+	var/lastangle = 0
+	var/aiming_lastangle = 0
+	var/aiming_time = 12
+	var/aiming_time_left = 12
+	var/laser_color = rgb(255,0,0)
+	var/listeningTo = null
+	var/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/last_beam
+	actions_list = list(/datum/action/item_action/toggle_laser_sight, /datum/action/item_action/change_laser_sight_color)
+
+/obj/item/attachment/laser_sight/examine(mob/user)
+	. = ..()
+	. += span_notice("You can <b>alt-click</b> it to change its color.")
+
+/obj/item/attachment/laser_sight/AltClick(mob/user)
+	. = ..()
+	var/C = input(user, "Select Laser Color", "Select laser color", laser_color) as null|color
+	if(!C || QDELETED(src))
+		return
+	laser_color = C
+
+/obj/item/attachment/laser_sight/on_attach(obj/item/gun/G, mob/user = null)
+	. = ..()
+	START_PROCESSING(SSfastprocess, src)
+	if(is_on)
+		G.spread -= 6
+
+/obj/item/attachment/laser_sight/on_detach(obj/item/gun/G, mob/living/user = null)
+	. = ..()
+	STOP_PROCESSING(SSfastprocess, src)
+	if(is_on)
+		G.spread += 6
+		QDEL_LIST(attached_gun.current_tracers)
+
+/obj/item/attachment/laser_sight/attack_self(mob/user)
+	. = ..()
+	toggle_on()
+
+/obj/item/attachment/laser_sight/proc/toggle_on()
+	is_on = !is_on
+	playsound(loc, is_on ? 'sound/weapons/magin.ogg' : 'sound/weapons/magout.ogg', 40, 1)
+	if(attached_gun)
+		if(is_on)
+			attached_gun.spread -= 6
+		else
+			attached_gun.spread += 6
+			QDEL_LIST(attached_gun.current_tracers)
+	update_icon()
+
+/obj/item/attachment/laser_sight/process()
+	return aiming_beam(TRUE)
+
+/obj/item/attachment/laser_sight/proc/aiming_beam(force_update = FALSE)
+	if(!is_on)
+		return
+	if(!attached_gun)
+		return PROCESS_KILL
+	if(!check_user())
+		// Doesn't need to be optimized, runs nothing if the list is already empty
+		QDEL_LIST(attached_gun.current_tracers)
+		return
+	process_aim()
+	var/diff = abs(aiming_lastangle - lastangle)
+	if(diff < 0.1 && !force_update)
+		return
+	aiming_lastangle = lastangle
+	var/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/P = new
```

Comment:
why p

Reply:
P for projectile, copied from beam rifle code